hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/capacity/QueuePath.java

```diff
@@ -116,8 +121,79 @@ public boolean hasParent() {
     return parent != null;
   }
 
+  /**
+   * Creates a new {@code QueuePath} from the current full path as parent, and
+   * the appended child queue path as leaf.
+   * @param childQueue path of leaf queue
+   * @return new queue path made of current full path and appended leaf path
+   */
+  public QueuePath append(String childQueue) {
```

Comment:
I think this append method is a bit misleading. It constructs a new QueuePath and does not do anything with "this", I think a constructor would be easier to understand:

QueuePath(QueuePath parent, String leaf) {}

(A static factory method would do the trick as well.)

Reply:
getFullPath is using this. This method is used for convenience, because a static constructor would be tedious. A new QueuePath is created here, because the whole class must maintain immutability, in order to be usable in HashSets and HashMaps.

Reply:
Lets wait for other reviews, I could change it if others find it confusing as well.

Reply:
I agree with @tomicooler, append indeed looks strange like this.